File: 0747-min-cost-climbing-stairs/0747-min-cost-climbing-stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int[] dp = new int[cost.length + 1];
        Arrays.fill(dp, -1);
        return Math.min(solve(cost, dp, 0), dp[1]);
    }

    int solve(int[] cost, int[] dp, int i){
        if(i >= cost.length){
            return 0;
        }
        if(dp[i] != -1)
            return dp[i];
        dp[i] = cost[i] + Math.min(solve(cost, dp, i+1), solve(cost, dp, i+2));
        return dp[i];
    }
}

File: 0747-min-cost-climbing-stairs/0747-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = [-1] * (len(cost)+1)
        
        def dfs(cost, i):
            if i >= len(cost):
                return 0
            if dp[i] != -1:
                return dp[i]
            dp[i] = cost[i] + min(dfs(cost, i+1), dfs(cost, i+2))
            return dp[i]

        res = dfs(cost, 0)
        return min(res, dp[1])